generator/doc/contributor/howto-guide-new-sidekick-codec.md

```diff
@@ -0,0 +1,122 @@
+more than one service (in the "gRPC service" sense in the same `api.API`
+invocation).
+
+Sidekick requires that all the elements in an `api.API` are in the same
+namespace. That is, all the elements in the generation lists must have the same
+`.Package` value in their field, and none may have the same `.ID` field.
+
+This is trivial for OpenAPI. For Protobuf this implies all the root messages,
+root enums, and services must be part of the same package. If you need to
+generate multiple packages that will require different calls to `Generate()`.
+
+To support mixins, `api.API` may reference messages, enums and and services.
+These are found in the `api.API.State.*ByID` maps.
+
+## How does sidekick gets its configuration?
+
+In a SDK repository the client libraries are generated in separate directories.
+Each directory contains a `.sidekick.toml` file. This file describes where to
+find the source for the client library, and may include some extra configuration
+for the codec. The top-level directory contains a `.sidekick.toml` configuration
+file which applies to all the generated client libraries.
+
+## How is `api.API` built?
+
+The `.sidekick.toml` file describes what parser to use for that directory, and
+where to find the source. The most common parser is `protobuf`. This parser
+typically receives a directory as the source, the parser uses all the `.proto`
+files in that directory (without recursion). It is also possible to list
+specific files, or to exclude some files.
+
+The parser calls `protoc`, reads the resulting proto descriptors and the service
+config YAML, and then converts that into a `api.API` instance. This is the input
+into your codec.
+
+## Implementing a `codec`
+
+First, add a module, something like this would do:
+
+```shell
+ls -l generator/internal/codec_sample
+```
+
+The main entry point in that module is the `Generate()` function, found in
+`generate.go`:
+
+```shell
+cat generator/internal/codec_sample/generate.go
+```
+
+This module has a single (and relatively simple) template file:
+
+```shell
+cat generator/internal/codec_sample/templats/readme/README.md.mustache
+```
+
+The codec is invoked from a single point in sidekick:
+
+```shell
+git grep -A 2 -B 2 codec_sample.Generate
+```
+
+## Basic integration tests
+
+A simple integration test for this module is found in:
+
+```shell
+cat generator/internal/sidekick/sidekick_sample_test.go
+```
+
+The output from this integration test is here:
+
+```shell
+cat generator/testdata/sample/golden/type/*
+```
+
+The CI builds will detect if these files changed unexpectedly. See [#1500] for
+ideas on how to improve these tests.
+
+## Unit tests
+
+You can write tests for the codec as usual. There are some helpers to initialize
+the `api.API` data structure. Look at the other codecs for examples.
+
+## Annotations
+
+Eventually your codec will need to annotate the `api.API` structure. Look at the
```

Comment:
Can you say what annotate means in this context?

Maybe "annotate the structure to add extra language-specific information" (if I know what annotate means)

Reply:
Done.